expo/expo
# [location] Split location permissions into background and foreground

# Why

This is a draft PR to comply with the Android 11 privacy changes. It splits up the `Permissions.LOCATION` into `Permissions.LOCATION_FOREGROUND` and `Permissions.LOCATION_BACKGROUND` to ["incrementally ask for location permissions"](https://developer.android.com/training/location/permissions#request-location-access-runtime).

- [User facing changes Android 10 vs 11](https://9to5google.com/2020/06/29/android-11-privacy/)
- [Short overview of the scenarios we need to support for Android 8, 9, 10, and 11](https://medium.com/swlh/request-location-permission-correctly-in-android-11-61afe95a11ad) _spoiler alert, they are all different_ 😢 

# How

**Updated `expo-permissions` to split up `LOCATION`**
- `Permissions.LOCATION` is the same as `Permissions.LOCATION_FOREGROUND`
- `Permissions.LOCATION_FOREGROUND` only asks for `ACCESS_FINE_LOCATION` and `ACCESS_COARSE_LOCATION` (through `LocationRequester`)
- `Permissions.LOCATION_BACKGROUND` only asks for `ACCESS_BACKGROUND_LOCATION` (through `SimpleRequester`)

> On Web, both `FOREGROUND` and `BACKGROUND` are handled by the same logic. Both of these permissions will just ask the location of the browser.

# Todo

- [ ] **iOS**: Refactor `expo-location` to support the split location permissions.
- [x] **Android**: Refactor `expo-location` and `expo-permissions` to support the split location permissions
- [x] **Web**: Refactor `expo-location` and `expo-permissions` to support the split location permissions

# Test Plan

- Make sure https://github.com/bycedric/office-marathon works with this change
- See if the current behavior of background location works on Android 8, 9, 10 _**and**_ 11.
- See if we can use foreground location only on Android 11
